snapcore/snapcraft
# lxd: friendly errror with suggestions if network is broken

-----
As evidenced by [LP: #1749679](https://bugs.launchpad.net/snapcraft/+bug/1749679) the verbatim network probing failures aren't very helpful. We can try to reduce noise and give some hints as to what's wrong. `ContainerConnectionError` will ensure there's a link to LXD docs and the URL we check is mentioned in the message.
`urllib.error.URLError` is handled explicitly to get a niceish error message in the case of connection errors. All other exceptions will generically be caught as well since we don't want a stack trace in any event and **urllib** can leak other exceptions such as `UnicodeError`.

Fixes: [LP: #1719952](https://bugs.launchpad.net/snapcraft/+bug/1719952)

The following tests were updated:

 - tests.unit.commands.test_snap
 - tests.unit.test_lxd

I locally ran the tests:
 - `./runtests.sh tests/unit` with unrelated failures in [tests.unit.test_lifecycle.CoreSetupTestCase.test_core_setup_if_docker_env](https://bugs.launchpad.net/snapcraft/+bug/1752576), `tests.unit.test_mangling.TestClearExecstack.test_execstack_clears` and `tests.unit.test_elf`.
 - `./runtests.sh tests/integration` with one unrelated failure in [tests.integration.general.test_parser.TestParserWikis](https://bugs.launchpad.net/snapcraft/+bug/1752580)
 - `./runtests.sh static`: Everything passed

Manual test steps:
 - cd tests/integration/snaps/basic
 - snapcraft cleanbuild
 - lxc profile set default environment.http_proxy localhost:1234
 - Observe `Failed to open 'http://start.ubuntu.com/connectivity-check.html': [Errno 111] Connection refused`
 - lxc profile set default environment.http_proxy invalid
 - Observe `Failed to open 'http://start.ubuntu.com/connectivity-check.html': [Errno -2] Name or service not known`
  - lxc profile set default environment.http_proxy http://unknown
  - Observe `Failed to open 'http://start.ubuntu.com/connectivity-check.html': [Errno -2] Name or service not known`
 - lxc profile set default environment.http_proxy .invalid
 - Observe `Failed to open 'http://start.ubuntu.com/connectivity-check.html': encoding with 'idna' codec failed (UnicodeError: label empty or too long)`

Note: Use `lxc profile set default environment.http_proxy ''` to unset the proxy in the default LXD profile after testing, or set it to the value you were using previously.
